### PR TITLE
Java: Update MaD Declarations after Triage

### PR DESCRIPTION
**Pull request by kaeluka**
This PR updates the MaD declarations after a recent triage session.

TR 2313

### CODE REVIEW

File: java/ql/lib/ext/com.google.common.io.model.yml [hunk 1]
```diff
@@ -3,6 +3,13 @@ extensions:
       pack: codeql/java-all
       extensible: sinkModel
     data:
+      - ["com.google.common.io", "Files", False, "asCharSink", "(File,Charset,FileWriteMode[])", "", "Argument[0]", "write-file", "ai-generated"] # this is a bit like a stream
```
Comment (atorralba) [inline java/ql/lib/ext/com.google.common.io.model.yml:6]: Usual counter-intuitive gotcha: `write-file` refers to the contents being written. So this is actually `create-file`:

```suggestion
      - ["com.google.common.io", "Files", False, "asCharSink", "(File,Charset,FileWriteMode[])", "", "Argument[0]", "create-file", "ai-generated"]
```

File: java/ql/lib/ext/com.google.common.io.model.yml [hunk 2]
```diff
@@ -3,6 +3,13 @@ extensions:
...
+      - ["com.google.common.io", "Files", False, "asCharSource", "(File,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["com.google.common.io", "Files", False, "copy", "(File,OutputStream)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["com.google.common.io", "Files", False, "readLines", "(File,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["com.google.common.io", "Files", False, "toByteArray", "(File)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["com.google.common.io", "Files", False, "toString", "(File,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["com.google.common.io", "Files", False, "write", "(byte[],File)", "", "Argument[1]", "write-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/com.google.common.io.model.yml:12]: Similarly, `write-file` should be `Argument[0]`, and we should also have a `create-file` sink for `Argument[1]`:

```suggestion
      - ["com.google.common.io", "Files", False, "write", "(byte[],File)", "", "Argument[0]", "write-file", "ai-generated"]
      - ["com.google.common.io", "Files", False, "write", "(byte[],File)", "", "Argument[1]", "create-file", "manual"]
```

File: java/ql/lib/ext/com.hippo.unifile.model.yml [hunk 3]
```diff
@@ -0,0 +1,8 @@
+extensions:
+  - addsTo:
+      pack: codeql/java-all
+      extensible: sinkModel
+    data:
+      - ["com.hippo.unifile", "UniFile", True, "createDirectory", "(String)", "", "Argument[0]", "create-file", "ai-generated"]
+      - ["com.hippo.unifile", "UniFile", True, "createFile", "(String)", "", "Argument[0]", "create-file", "ai-generated"]
+      - ["com.hippo.unifile", "UniFile", True, "renameTo", "(String)", "", "Argument[0]", "create-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/com.hippo.unifile.model.yml:8]: These depend on the implementation (they're abstract), and I could find at least one vulnerable implementation and two safe implementations. Also, the library doesn't seem to be too popular (judging by its stars/forks in GitHub), so maybe we should leave this whole file out.

File: java/ql/lib/ext/java.lang.model.yml [hunk 4]
```diff
@@ -4,8 +4,15 @@ extensions:
...
+# suggested label is not supported:      - ["java.lang", "ProcessBuilder", False, "command", "(String[])", "", "Argument[0]", "command-injection", "ai-generated"]
+# suggested label is not supported:      - ["java.lang", "ProcessBuilder", False, "ProcessBuilder", "(List)", "", "Argument[0]", "command-injection", "ai-generated"]
+# suggested label is not supported:      - ["java.lang", "Runtime", True, "exec", "(String,String[],File)", "", "Argument[0]", "command-injection", "ai-generated"]
+# suggested label is not supported:      - ["java.lang", "Runtime", True, "exec", "(String,String[])", "", "Argument[0]", "command-injection", "ai-generated"]
+# suggested label is not supported:      - ["java.lang", "Runtime", True, "exec", "(String[],String[])", "", "Argument[0]", "command-injection", "ai-generated"]
       # These are modeled in plain CodeQL. TODO: migrate them.
       # - ["java.lang", "ProcessBuilder", False, "directory", "(File)", "", "Argument[0]", "command-injection", "ai-generated"]
       # - ["java.lang", "ProcessBuilder", True, "ProcessBuilder", "(String[])", "", "Argument[0]", "command-injection", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.lang.model.yml:18]:
```suggestion
      # These are modeled in plain CodeQL. TODO: migrate them.
      # - ["java.lang", "ProcessBuilder", False, "command", "(String[])", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "ProcessBuilder", False, "directory", "(File)", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "ProcessBuilder", False, "ProcessBuilder", "(List)", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "ProcessBuilder", False, "ProcessBuilder", "(String[])", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "Runtime", True, "exec", "(String,String[])", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "Runtime", True, "exec", "(String[],String[])", "", "Argument[0]", "command-injection", "ai-generated"]
      # - ["java.lang", "Runtime", True, "exec", "(String,String[],File)", "", "Argument[0]", "command-injection", "ai-generated"]
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 5]
```diff
@@ -11,14 +11,26 @@ extensions:
       - ["java.nio.file", "Files", False, "createLink", "", "", "Argument[0]", "create-file", "manual"]
       - ["java.nio.file", "Files", False, "createSymbolicLink", "", "", "Argument[0]", "create-file", "manual"]
       - ["java.nio.file", "Files", False, "createTempDirectory", "(Path,String,FileAttribute[])", "", "Argument[0]", "create-file", "manual"]
+      - ["java.nio.file", "Files", False, "createTempDirectory", "(String,FileAttribute[])", "", "Argument[0]", "create-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:14]: Controlling the prefix doesn't let you control the actual directory the file is created in.

```suggestion
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 6]
```diff
@@ -11,14 +11,26 @@ extensions:
...
+      - ["java.nio.file", "Files", False, "createTempDirectory", "(String)", "", "Argument[0]", "create-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:15]: Does this signature even exist?

```suggestion
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 7]
```diff
@@ -11,14 +11,26 @@ extensions:
...
       - ["java.nio.file", "Files", False, "createTempFile", "(Path,String,String,FileAttribute[])", "", "Argument[0]", "create-file", "manual"]
+      - ["java.nio.file", "Files", False, "createTempFile", "(Path,String,String)", "", "Argument[0]", "create-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:17]: Same, this doesn't exist. Seems to be a confusion with the variadic arguments.

```suggestion
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 8]
```diff
@@ -11,14 +11,26 @@ extensions:
...
       - ["java.nio.file", "Files", False, "delete", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
       - ["java.nio.file", "Files", False, "deleteIfExists", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
+# suggested label is not supported:      - ["java.nio.file", "Files", False, "deleteIfExists", "(Path)", "", "Argument[0]", "delete-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:20]:
```suggestion
      - ["java.nio.file", "Files", False, "deleteIfExists", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 9]
```diff
@@ -11,14 +11,26 @@ extensions:
...
       - ["java.nio.file", "Files", False, "lines", "(Path,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["java.nio.file", "Files", False, "lines", "(Path,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:22]: Duplicates the model just above:

```suggestion
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 10]
```diff
@@ -11,14 +11,26 @@ extensions:
...
+      - ["java.nio.file", "Files", False, "lines", "(Path)", "", "Argument[0]", "read-file", "ai-generated"]
       - ["java.nio.file", "Files", False, "move", "", "", "Argument[1]", "create-file", "manual"]
       - ["java.nio.file", "Files", False, "newBufferedReader", "(Path,Charset)", "", "Argument[0]", "read-file", "ai-generated"]
+      - ["java.nio.file", "Files", False, "newBufferedReader", "(Path)", "", "Argument[0]", "read-file", "ai-generated"]
       - ["java.nio.file", "Files", False, "newBufferedWriter", "", "", "Argument[0]", "create-file", "manual"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:27]: These three overlap, let's keep the manual one:

```suggestion
      - ["java.nio.file", "Files", False, "newBufferedWriter", "", "", "Argument[0]", "create-file", "manual"]
```

File: java/ql/lib/ext/java.nio.file.model.yml [hunk 11]
```diff
@@ -28,6 +40,8 @@ extensions:
       - ["java.nio.file", "Files", True, "delete", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
       - ["java.nio.file", "Files", True, "newInputStream", "(Path,OpenOption[])", "", "Argument[0]", "read-file", "ai-generated"]
       - ["java.nio.file", "Files", True, "newOutputStream", "(Path,OpenOption[])", "", "Argument[0]", "create-file", "ai-generated"]
+# suggested label is not supported:      - ["java.nio.file", "SecureDirectoryStream", True, "deleteDirectory", "(Path)", "", "Argument[0]", "delete-file", "ai-generated"]
+# suggested label is not supported:      - ["java.nio.file", "SecureDirectoryStream", True, "deleteFile", "(Path)", "", "Argument[0]", "delete-file", "ai-generated"]
```
Comment (atorralba) [inline java/ql/lib/ext/java.nio.file.model.yml:44]:
```suggestion
      - ["java.nio.file", "SecureDirectoryStream", True, "deleteDirectory", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
      - ["java.nio.file", "SecureDirectoryStream", True, "deleteFile", "(Path)", "", "Argument[0]", "create-file", "ai-generated"] # should be delete-file
```

**Review by atorralba — COMMENTED** [review]
Thanks for this, @kaeluka. I added some suggestions, let me know if anything doesn't make sense to you. Otherwise I'll apply them and merge 👍

**Comment by kaeluka** [comment]
> Thanks for this, @kaeluka. I added some suggestions, let me know if anything doesn't make sense to you. Otherwise I'll apply them and merge 👍

I've looked over the suggestions and accepted them! Thx :)

The question about `com.hippo.unifile.model.yml` is still open — I personally don't have much of an opinion. If you want to drop the file, go for it :) Cases like this should be fixed in by using the sort utility next time :)

**Comment by kaeluka** [comment]
Feel free to merge any time 👍

**Review by atorralba — APPROVED** [review]

**Pull request merged**